File: android/app/src/main/java/com/example/website_blocker/VpnService.java
// android/app/src/main/java/com/example/website_blocker/VpnService.java
package com.example.website_blocker;

import android.content.Intent;
import android.net.VpnService;
import android.os.ParcelFileDescriptor;
import android.util.Log;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.util.Set;
import java.util.HashSet;

public class VpnService extends android.net.VpnService {
    private static final String TAG = "WebsiteBlockerVpnService";
    private Thread mThread;
    private ParcelFileDescriptor mInterface;
    private boolean running = false;
    private Set<String> blockedDomains = new HashSet<>();
    
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // Retrieve blocked domains from the intent if provided
        if (intent != null && intent.hasExtra("blockedDomains")) {
            String[] domains = intent.getStringArrayExtra("blockedDomains");
            if (domains != null) {
                for (String domain : domains) {
                    blockedDomains.add(domain);
                }
            }
        }
        
        startVpn();
        return START_STICKY;
    }
    
    private void startVpn() {
        if (mThread != null) {
            mThread.interrupt();
        }
        
        running = true;
        
        mThread = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    runVpn();
                } catch (Exception e) {
                    Log.e(TAG, "VPN service error", e);
                } finally {
                    stopVpn();
                }
            }
        }, "WebsiteBlockerVpnThread");

        // android/app/src/main/java/com/example/website_blocker/VpnService.java (continued)
        mThread.start();
    }
    
    private void runVpn() throws Exception {
        // Configure the VPN
        Builder builder = new Builder();
        builder.setSession("Website Blocker VPN")
               .addAddress("10.0.0.2", 32)
               .addRoute("0.0.0.0", 0)
               .addDnsServer("8.8.8.8")
               .setMtu(1500);
       
        // Create the interface
        mInterface = builder.establish();
        if (mInterface == null) {
            throw new IllegalStateException("Failed to establish VPN connection");
        }
        
        // Process packets
        processPackets();
    }
    
    private void processPackets() throws Exception {
        FileInputStream in = new FileInputStream(mInterface.getFileDescriptor());
        FileOutputStream out = new FileOutputStream(mInterface.getFileDescriptor());
        ByteBuffer packet = ByteBuffer.allocate(32767);
        
        // Create a socket for sending packets
        DatagramChannel tunnel = DatagramChannel.open();
        tunnel.connect(new InetSocketAddress("127.0.0.1", 8087));
        
        while (running) {
            // Read the outgoing packet from the input stream
            int length = in.read(packet.array());
            if (length > 0) {
                // Packet analysis would happen here
                // We'd analyze the packet, check if it's HTTP/HTTPS and if the domain
                // matches one of our blocked domains
                
                // For a complete implementation, we would:
                // 1. Parse the IP/TCP headers
                // 2. Extract the domain from HTTP/SNI if it's HTTPS
                // 3. Check if the domain is in our blocklist
                // 4. Either forward the packet or block it

                // This is a simplified implementation
                boolean shouldBlock = analyzePacket(packet.array(), length);
                
                if (!shouldBlock) {
                    // Write the packet to the tunnel
                    packet.limit(length);
                    tunnel.write(packet);
                    packet.clear();
                    
                    // Read the response
                    length = tunnel.read(packet);
                    if (length > 0) {
                        out.write(packet.array(), 0, length);
                    }
                    packet.clear();
                } else {
                    // Packet is blocked, do not forward it
                    // For HTTP requests, we could optionally return a custom response
                    Log.d(TAG, "Blocking packet to blocked domain");
                }
            }
        }
    }
    
    private boolean analyzePacket(byte[] packet, int length) {
        // This is a simplified version - in a real implementation, 
        // you'd parse the packet and extract the domain
        // For the purpose of this example, we'll assume it always passes
        
        // TODO: Implement proper packet analysis to extract domains
        return false;
    }
    
    private void stopVpn() {
        running = false;
        if (mInterface != null) {
            try {
                mInterface.close();
                mInterface = null;
            } catch (Exception e) {
                Log.e(TAG, "Error closing VPN interface", e);
            }
        }
    }
    
    @Override
    public void onDestroy() {
        if (mThread != null) {
            mThread.interrupt();
        }
        stopVpn();
        super.onDestroy();
    }
    
    public void updateBlockedDomains(Set<String> domains) {
        blockedDomains.clear();
        blockedDomains.addAll(domains);
    }
}

File: android/app/src/main/kotlin/com/example/website_blocker/VpnService.kt
package com.example.website_blocker

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.VpnService
import android.os.Build
import android.os.ParcelFileDescriptor
import android.system.OsConstants
import androidx.core.app.NotificationCompat
import java.io.FileInputStream
import java.io.FileOutputStream
import java.net.InetAddress
import java.util.concurrent.atomic.AtomicBoolean

class WebsiteBlockerVpnService : VpnService() {
    private var vpnInterface: ParcelFileDescriptor? = null
    private val isRunning = AtomicBoolean(false)
    private var blockedDomains: Set<String> = emptySet()

    companion object {
        private const val NOTIFICATION_ID = 1
        private const val CHANNEL_ID = "vpn_service_channel"
        private const val CHANNEL_NAME = "VPN Service"
    }

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_DEFAULT
            )
            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun createNotification(): Notification {
        val intent = Intent(this, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(
            this, 0, intent,
            PendingIntent.FLAG_IMMUTABLE
        )

        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("Website Blocker VPN")
            .setContentText("VPN is active")
            .setSmallIcon(R.drawable.ic_launcher)
            .setContentIntent(pendingIntent)
            .build()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        try {
            when (intent?.action) {
                "start" -> {
                    val domains = intent.getStringArrayListExtra("blockedDomains")?.toSet() ?: emptySet()
                    startVpn(domains)
                }
                "stop" -> stopVpn()
                "update" -> {
                    val domains = intent.getStringArrayListExtra("blockedDomains")?.toSet() ?: emptySet()
                    updateBlockedDomains(domains)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            stopSelf()
        }
        return START_STICKY
    }

    private fun startVpn(domains: Set<String>) {
        if (isRunning.get()) return

        try {
            val builder = VpnService.Builder()
                .setSession("Website Blocker VPN")
                .addAddress("10.0.0.2", 32)
                .addDnsServer("8.8.8.8")
                .addDnsServer("8.8.4.4")
                .addRoute("0.0.0.0", 0)

            vpnInterface = builder.establish()
            blockedDomains = domains
            isRunning.set(true)

            startForeground(NOTIFICATION_ID, createNotification())
            Thread { runVpn() }.start()
        } catch (e: Exception) {
            e.printStackTrace()
            stopSelf()
        }
    }

    private fun stopVpn() {
        try {
            isRunning.set(false)
            vpnInterface?.close()
            vpnInterface = null
            stopForeground(true)
            stopSelf()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun updateBlockedDomains(domains: Set<String>) {
        blockedDomains = domains
    }

    private fun runVpn() {
        try {
            val input = FileInputStream(vpnInterface?.fileDescriptor)
            val output = FileOutputStream(vpnInterface?.fileDescriptor)
            val buffer = ByteArray(32767)

            while (isRunning.get()) {
                val length = input.read(buffer)
                if (length > 0) {
                    if (!shouldBlockPacket(buffer, length)) {
                        output.write(buffer, 0, length)
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            stopSelf()
        }
    }

    private fun shouldBlockPacket(packet: ByteArray, length: Int): Boolean {
        if (length < 12) return false

        try {
            // Check if it's a DNS query
            if (packet[2] and 0x80.toByte() == 0.toByte()) {
                val domain = extractDomain(packet, length)
                return blockedDomains.any { domain.contains(it) }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return false
    }

    private fun extractDomain(packet: ByteArray, length: Int): String {
        var domain = ""
        var offset = 12 // Skip DNS header

        try {
            while (offset < length && packet[offset] != 0.toByte()) {
                val labelLength = packet[offset].toInt() and 0xFF
                offset++

                if (offset + labelLength > length) break

                val label = String(packet, offset, labelLength)
                domain += label + "."
                offset += labelLength
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return domain
    }

    override fun onDestroy() {
        super.onDestroy()
        stopVpn()
    }
} 